projecttacoma/measure-repository-service-test-kit
# Include terminology parameter optional tests

# Summary
Functionality for an `include-terminology` parameter in `$package` operations in  the measure-repository-service was added with [this PR](https://github.com/projecttacoma/measure-repository-service/pull/11). This PR adds an optional test to test that when this flag is provided in `$package` operations (both Measure and Library), the resulting bundle includes valueset resources (it shouldn't by default).

## New behavior
- ~Two new test groups: one for the `Measure/$package` operation with `include-terminology=true` query parameter and one for the `Library/$package` operation with `include-terminology=true` query parameter.~
- rspec tests for both groups.
- Functions for getting the related artifacts from a bundle that are valuesets.

**EDIT:**
- New test in each of the existing `Measure/$package` and `Library/$package` test groups with `include-terminology=true` query parameter.

## Code changes
- ~`measure_repository_service_test_kit.rb` - added the two new test groups~
- ~`library_include_terminology.rb` - test group to test the `Library/$package` operation with `include-terminology=true`. All of the tests are the same as the regular `Library/$package` test group with the exception of making sure the valuesets are now including. Let me know if any of these tests are redundant.~
- ~`measure_include_terminology.rb` - test group to test the `Measure/$package` operation with `include-terminology=true`. All of the tests are the same as the regular `Measure/$package` test group with the exception of making sure the valuesets are now including. Let me know if any of these tests are redundant.~
- `package_utils.rb` - 3 new functions that are similar to the ones used to find the related library artifacts but these find the related valueset artifacts.
- ~`library_include_terminology_spec.rb` / `measure_include_terminology_spec.rb` - spec tests. Basically the same as the spec tests for the regular Measure and Library `$package` test groups.~

**EDIT:**
- Rather than making two new test groups, I added a test for the `include-terminology=true` functionality to each existing test group (`measure_package.rb` and `library_package.rb`.)
- Corresponding rspec tests were added to `library_package_spec.rb`and `measure_package_spec.rb`.

# Testing guidance
- `rspec` and `rubocop`
- in `measure-respository-service` (make sure it's up to date):
```
git clone https://github.com/cqframework/ecqm-content-r4-2021.git
npm run db:reset
npm run db:loadBundle "ecqm-content-r4-2021/bundles/measure/ColorectalCancerScreeningsFHIR/ColorectalCancerScreeningsFHIR-bundle.json"
```
- Start up `measure-repository-service` with `npm start`
- Follow the directions in the `README.md` for local setup for the test kit
- Navigate to `http://localhost:4567`
- You should see that there are still 5 test groups, but that `Measure/$package` and  `Library/$package` each have an additional _optional_ test.
- Click "RUN ALL TESTS"
- Enter the following input for the respective fields:
```
FHIR Server Base Url (required) *: http://localhost:3000/4_0_1
selected_measure_id (required) *: ColorectalCancerScreeningsFHIR
selected_library_id (required) *: ColorectalCancerScreeningsFHIR
selected_measure_url (required) *: http://ecqi.healthit.gov/ecqms/Measure/ColorectalCancerScreeningsFHIR
selected_measure_identifier (required) *: http://hl7.org/fhir/cqi/ecqm/Measure/Identifier/cms|130FHIR
selected_measure_version:  0.0.003
selected_library_url (required) *: http://ecqi.healthit.gov/ecqms/Library/ColorectalCancerScreeningsFHIR
selected_library_identifier (required) *: http://hl7.org/fhir/cqi/ecqm/Measure/Identifier/cms|130FHIR
selected_library_version:  0.0.003
```
- Click submit and watch all the tests pass **except** for two in the`Library/$package` test suite. This is expected! The EXM130 measure does not have any identifier's in its libraries so the test with identifier should fail. 
- The new optional tests should pass.
- Try the same except start up the test kit using docker, and change FHIR Server Base Url to `host.docker.internal:3000/4_0_1` and everything should still work.
- ~I just initially made these two test groups separate, but should they be apart of their respective $package operation test group? Let me know how you feel about the structure.~ Changed!

Notes:
- Might be helpful to look at [this Jira ticket](https://jira.hl7.org/browse/FHIR-40219) and the PR that implements this in the measure-repository-service (linked above).
